event/datadeal/data_delivery_event.go

```diff
@@ -0,0 +1,235 @@
+	reactor event.Reactor
+}
+
+func NewDataDeliveryVoteEvent(s event.Reactor) DataDeliveryVoteEvent {
+	return DataDeliveryVoteEvent{s}
+}
+
+func (e DataDeliveryVoteEvent) GetEventType() string {
+	return "data_delivery"
+}
+
+func (e DataDeliveryVoteEvent) GetEventAttributeKey() string {
+	return "vote_status"
+}
+
+func (e DataDeliveryVoteEvent) GetEventAttributeValue() string {
+	return "'started'"
+}
+
+func (e DataDeliveryVoteEvent) EventHandler(event ctypes.ResultEvent) error {
+
+	dealIDStr := event.Events[datadealtypes.EventTypeDataDeliveryVote+"."+datadealtypes.AttributeKeyDealID][0]
+	dataHash := event.Events[datadealtypes.EventTypeDataDeliveryVote+"."+datadealtypes.AttributeKeyDataHash][0]
+
+	dealID, err := strconv.ParseUint(dealIDStr, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	oraclePrivKey := e.reactor.OraclePrivKey()
+
+	voteOption, deliveredCid, err := e.verifyAndGetVoteOption(dealID, dataHash, oraclePrivKey)
+	if err != nil {
+		log.Infof("error while verify: %v", err)
+		if voteOption == oracletypes.VOTE_OPTION_UNSPECIFIED {
+			return err
+		}
+	}
+
+	msgVoteDataDelivery, err := e.makeDataDeliveryVote(
+		e.reactor.OracleAcc().GetAddress(),
+		dataHash,
+		deliveredCid,
+		dealID,
+		voteOption,
+		oraclePrivKey.Serialize(),
+	)
+	if err != nil {
+		return err
+	}
+
+	txBuilder := panacea.NewTxBuilder(*e.reactor.QueryClient())
+
+	txBytes, err := e.generateTxBytes(msgVoteDataDelivery, e.reactor.OracleAcc().GetPrivKey(), e.reactor.Config(), txBuilder)
+	if err != nil {
+		return err
+	}
+
+	if err := e.broadcastTx(e.reactor.GRPCClient(), txBytes); err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+func (e DataDeliveryVoteEvent) verifyAndGetVoteOption(dealID uint64, dataHash string, oraclePrivKey *btcec.PrivateKey) (oracletypes.VoteOption, string, error) {
+
+	dataSale, err := e.reactor.QueryClient().GetDataSale(dealID, dataHash)
+	if err != nil {
+		if err == datadealtypes.ErrDataSaleNotFound {
+			return oracletypes.VOTE_OPTION_NO, "", err
+		} else {
+			return oracletypes.VOTE_OPTION_UNSPECIFIED, "", err
+		}
+	}
+
+	if dataSale.Status != datadealtypes.DATA_SALE_STATUS_DELIVERY_VOTING_PERIOD {
+		return oracletypes.VOTE_OPTION_NO, "", nil
+	}
+
+	if len(dataSale.VerifiableCid) == 0 {
+		return oracletypes.VOTE_OPTION_NO, "", nil
+	}
+
+	deal, err := e.reactor.QueryClient().GetDeal(dataSale.DealId)
+	if err != nil {
+		if err == datadealtypes.ErrDealNotFound {
+			return oracletypes.VOTE_OPTION_NO, "", err
+		} else {
+			return oracletypes.VOTE_OPTION_UNSPECIFIED, "", err
+		}
+	}
+
+	deliveredCID, err := e.makeDeliveredCid(deal, dataSale, oraclePrivKey)
+	if err != nil {
+		return oracletypes.VOTE_OPTION_NO, "", fmt.Errorf("error while make deliveredCid: %v", err)
+	}
+
+	return oracletypes.VOTE_OPTION_YES, deliveredCID, nil
+
+}
+
+func (e DataDeliveryVoteEvent) makeDeliveredCid(deal *datadealtypes.Deal, dataSale *datadealtypes.DataSale, oraclePrivKey *btcec.PrivateKey) (string, error) {
```

Comment:
This function is to upload data that the buyer can decrypt to ipfs.
So I think it would be good to change function name to `convertBuyerDataAndUploadIpfs`
Well. Other names are possible, but I can't remember right now. :)

Reply:
Okay, I change this function name to `convertBuyerDataAndAddToIpfs`, because `upload` is expressed as `add` in ipfs.